Sitka-Sound-Science-Center/sitka-landslide
# Add script to fetch and format rainfall data

## Overview

This is a node script that can be run from the command line that fetches observed rainfall data from the MesoWest API and predicted future rainfall amounts from the NWS API, extracts the bits we care about, and applies the model to assign a risk level. It produces a JSON object, which it currently just prints the results to the console.

For observed data, it uses the [MesoWest `precip` endpoint](https://developers.synopticdata.com/mesonet/v2/stations/precipitation/).  I initially tried to use the [`timeseries endpoint`](https://developers.synopticdata.com/mesonet/v2/stations/timeseries/), but `precip` is much nicer for our purposes.  The precip measurements within `timeseries` include a lot of nulls--seemingly because it returns a bunch of different measurements in arrays, so there has to be an entry for every timestamp, even if there wasn't a precip measurement at that timestamp.  It also doesn't seem to distinguish between actual observations with zero precip and observations that didn't include a precip measurement at all.  The `precip` endpoint is more complex (it has different time and subtotal modes that are controlled by parameters) but it's able to produce the quantities we need in a way that we can just grab them, rather than having to filter and sum observations.

For forecasts, it uses [the NWS API `/gridpoints/{wfo}/{x},{y}`](https://www.weather.gov/documentation/services-web-api#/default/gridpoint) endpoint.  Here again, I started out looking at a different endpoint (`/gridpoints/{wfo}/{x},{y}/forecast`), but in this case that didn't go very far because it turns out that doesn't include precip forecasts at all.  The base `gridpoints` endpoint includes a _lot_ of information, most of which we don't need, but one of the quantities is `quantitativePrecipitation`, which provides precipitation amount forecasts, in millimeters, for each 3-hour period within the validity window of the forecast.

In the meeting with Annette and Jacyn on 3/3/22 about the details of how to apply the model to produce risk ratings, we decided to use a 6-hour buffer for assigning risk levels.  So the risk level for future time periods is calculated based on the highest total from the 3hr forecast window and the two 3hr windows before it.  The calculation for current risk based on observed rainfall is slightly different, because the endpoint returns the trailing rainfall totals, not fixed 3hr windows, so we can just use the higher of the past 3 hours or the 3 hours before that.  We haven't entirely ironed out how we want to present that nuance in the front-end, but I included both `precip` and `riskPrecip` in the results, so we'll be able to tell if the total used to calculate risk is not the latest total.

Resolves #2 and #3

### Demo

```json
{
  "observed": {
    "timestamp": "2022-03-08T11:53:00.000-09:00",
    "precip": 0,
    "riskPrecip": 0,
    "risk": "Low"
  },
  "forecast": [
    {
      "timestamp": "2022-03-08T12:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-08T15:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-08T18:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-08T21:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T00:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T03:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T06:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T09:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T12:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T15:00:00.000-09:00",
      "precip": 0,
      "riskPrecip": 0,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T18:00:00.000-09:00",
      "precip": 0.762,
      "riskPrecip": 0.762,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-09T21:00:00.000-09:00",
      "precip": 1.27,
      "riskPrecip": 1.27,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T00:00:00.000-09:00",
      "precip": 1.016,
      "riskPrecip": 1.27,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T03:00:00.000-09:00",
      "precip": 2.032,
      "riskPrecip": 2.032,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T06:00:00.000-09:00",
      "precip": 2.286,
      "riskPrecip": 2.286,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T09:00:00.000-09:00",
      "precip": 2.54,
      "riskPrecip": 2.54,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T12:00:00.000-09:00",
      "precip": 2.286,
      "riskPrecip": 2.54,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T15:00:00.000-09:00",
      "precip": 3.048,
      "riskPrecip": 3.048,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T18:00:00.000-09:00",
      "precip": 3.048,
      "riskPrecip": 3.048,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-10T21:00:00.000-09:00",
      "precip": 4.064,
      "riskPrecip": 4.064,
      "risk": "Low"
    },
    {
      "timestamp": "2022-03-11T00:00:00.000-09:00",
      "precip": 3.81,
      "riskPrecip": 4.064,
      "risk": "Low"
    }
  ]
}
```

### Notes

- The MesoWest API is free but requires an API key.  I registered for an account using my Azavea email address, and just committed the key to the repo.  We can get fancier than that later if necessary, but it didn't seem like enough of a sensitive secret to go to the trouble right now.
- The NWS API doesn't require an account or key at all, but they want you to included a User-Agent header identifying yourself, so they can get in touch if there's a problem.  It seems extremely unlikely that this app will attract any attention, but I included a name and an email address (systems@azavea.com) as requested.
- This finds the config files in the project root, but I didn't integrate actually install Prettier or eslint, or integrate this into the STRTA in any way.  I'm not sure exactly how this will be connected, so it seemed premature to try to integrate it too much.

## Testing Instructions

```
cd rainfall
nvm use && yvm use
yarn install
node --experimental-modules rainfall.js
```
That should print some JSON that looks similar to the example above.
If the forecast values are moving around, you should be able to see that `precip` moves up and down from one entry to the next but if `precip` is lower in one entry than it was in the one before, `riskPrecip` stays at the higher value until it's more than 6 hours old.

## Checklist

- [ ] ~`./scripts/lint` runs without errors~ (no linting set up here yet, though I did run Prettier in my editor)
